src/mlpack/core/optimizers/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -0,0 +1,141 @@
+ParallelSGD<SparseFunctionType, DecayPolicyType>::ParallelSGD(
+    SparseFunctionType& function,
+    const size_t maxIterations,
+    const size_t batchSize,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy) :
+    function(function),
+    maxIterations(maxIterations),
+    batchSize(batchSize),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy)
+{ /* Nothing to do. */ }
+
+template <typename SparseFunctionType, typename DecayPolicyType>
+double ParallelSGD<SparseFunctionType, DecayPolicyType>::Optimize(
+    SparseFunctionType& function,
+    arma::mat& iterate)
+{
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  for (size_t i = 1; i <= maxIterations; ++i){
+    overallObjective = 0;
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    arma::Col<size_t> visitationOrder;
+    GenerateVisitationOrder(visitationOrder);
+
+    #pragma omp parallel
+    {
+      // Each processor gets a subset of the instances.
+      // Each subset is of size batchSize.
+      arma::Col<size_t> instances = ThreadShare(omp_get_thread_num(),
+          visitationOrder);
+      for (size_t j = 0; j < instances.n_elem; ++j)
+      {
+        // Each instance affects only some components of the decision variable.
+        // So the gradient is sparse.
+        arma::sp_mat gradient;
+
+        // Evaluate the sparse gradient.
+        function.Gradient(iterate, instances[j], gradient);
+
+        // Update the decision variable with non-zero components of the
+        // gradient.
+        for(size_t i = 0; i < gradient.n_cols; ++i)
+        {
+          // Iterate over the non-zero elements.
+          for (auto cur = gradient.begin_col(i); cur != gradient.end_col(i);
+              ++cur)
+          {
+            #pragma omp atomic
+            iterate(cur.row(), i) -= stepSize * gradient(cur.row(), i);
+          }
+        }
+      }
+    }
+
+    // Evaluate the function
+    overallObjective = 0;
+    for (size_t j = 0; j < function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    Log::Info << "\nObjective : " << overallObjective << " Iteration : " << i;
+    if (std::abs(overallObjective - lastObjective) < tolerance)
+    {
+      Log::Info << "\nParallel SGD terminated with objective delta "
+        << " within tolerance : " << overallObjective << std::endl;
+      return overallObjective;
+    }
+    lastObjective = overallObjective;
+  }
+  Log::Info << "\n Parallel SGD terminated with objective : "
+    << overallObjective << std::endl;
+  return overallObjective;
+}
+
+template <typename SparseFunctionType, typename DecayPolicyType>
+void ParallelSGD<
+    SparseFunctionType,
+    DecayPolicyType>::GenerateVisitationOrder(
+        arma::Col<size_t>& visitationOrder)
+{
+  visitationOrder = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        (function.NumFunctions() - 1), function.NumFunctions()));
```

Comment:
Another idea that might be better for data locality is to split the points that are worked on before shuffling: so thread 0 gets [0, 10), thread 1 gets [10, 20), and so forth.  Then shuffle those individual sets for each thread.  I think this will be faster and won't have a significant effect on convergence, but I'm not 100% certain.